File: 1475-final-prices-with-a-special-discount-in-a-shop/1475-final-prices-with-a-special-discount-in-a-shop.java
class Solution {
    public int[] finalPrices(int[] prices) {
        for (int i = 0; i < prices.length; i++) {
            int pointer = i + 1;    
            while(pointer < prices.length) {
                if (prices[i] >= prices[pointer]) {
                    prices[i] -= prices[pointer];  
                    break;
                } 
                pointer++;
            }
        }
        return prices;
    }
}

File: 1475-final-prices-with-a-special-discount-in-a-shop/1475-final-prices-with-a-special-discount-in-a-shop.kt
class Solution {
    fun finalPrices(prices: IntArray): IntArray {
        val stack = ArrayDeque<Int>()

        for (i in 0..prices.size - 1) {
            while (stack.isNotEmpty() && prices[stack.last()] >= prices[i]) {
                prices[stack.removeLast()] -= prices[i]
            }
            stack.addLast(i)
        }

        return prices
    }
}